File: aib-backend/src/main/java/com/aib/aib_backend/service/ProductService.java
package com.aib.aib_backend.service;

import com.aib.aib_backend.dto.request.ProductRequest;
import com.aib.aib_backend.dto.request.ProductSearchRequest;
import com.aib.aib_backend.dto.response.ProductListResponse;
import com.aib.aib_backend.dto.response.ProductResponse;
import java.math.BigDecimal;
import java.util.List;

/**
 * Service interface for Product operations
 */
public interface ProductService {

    // ============================================
    // ADMIN OPERATIONS
    // ============================================

    /**
     * Create a new product
     * @param request Product details
     * @return Created product
     */
    ProductResponse createProduct(ProductRequest request);

    /**
     * Update existing product
     * @param id Product ID
     * @param request Updated product details
     * @return Updated product
     */
    ProductResponse updateProduct(Long id, ProductRequest request);

    /**
     * Delete product
     * @param id Product ID
     */
    void deleteProduct(Long id);

    /**
     * Toggle product active/inactive status
     * @param id Product ID
     * @return Updated product
     */
    ProductResponse toggleProductStatus(Long id);

    /**
     * Update product stock quantity
     * @param id Product ID
     * @param quantity New stock quantity
     * @return Updated product
     */
    ProductResponse updateStock(Long id, Integer quantity);

    // ============================================
    // PUBLIC OPERATIONS
    // ============================================

    /**
     * Search products with advanced filtering
     * @param searchRequest Search and filter parameters
     * @return Paginated product list with filter summary
     */
    ProductListResponse searchProducts(ProductSearchRequest searchRequest);

    /**
     * Get product by ID
     * @param id Product ID
     * @return Product details
     */
    ProductResponse getProductById(Long id);

    /**
     * Get product by slug
     * @param slug Product slug
     * @return Product details
     */
    ProductResponse getProductBySlug(String slug);

    /**
     * Get products by category with pagination
     * @param categoryId Category ID
     * @param page Page number
     * @param size Page size
     * @return List of products
     */
    List<ProductResponse> getProductsByCategory(Long categoryId, int page, int size);

    /**
     * Get featured products
     * @param limit Maximum number of products to return
     * @return List of featured products
     */
    List<ProductResponse> getFeaturedProducts(int limit);

    /**
     * Get discounted products
     * @param limit Maximum number of products to return
     * @return List of discounted products
     */
    List<ProductResponse> getDiscountedProducts(int limit);

    // ============================================
    // UTILITY METHODS
    // ============================================

    /**
     * Get all distinct brands
     * @return List of brand names
     */
    List<String> getAllBrands();




    /**
     * Get price range (min and max prices)
     * @return Price range
     */
    PriceRange getPriceRange(ProductSearchRequest searchRequest);
    /**
     * Price range wrapper class
     */
    class PriceRange {
        public BigDecimal minPrice;
        public BigDecimal maxPrice;

        public PriceRange(BigDecimal minPrice, BigDecimal maxPrice) {
            this.minPrice = minPrice;
            this.maxPrice = maxPrice;
        }
    }
}

File: aib-frontend/src/app/core/services/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Product, ProductListResponse, ProductRequest, ProductSearchRequest, PriceRange } from '@shared/models/product.model';
import { environment } from '@environments/environment';

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  private apiUrl = `${environment.apiUrl}/products`;

  constructor(private http: HttpClient) {}

  searchProducts(searchRequest: ProductSearchRequest): Observable<ProductListResponse> {
    return this.http.post<ProductListResponse>(`${this.apiUrl}/search`, searchRequest);
  }

  getProductById(id: number): Observable<Product> {
    return this.http.get<Product>(`${this.apiUrl}/${id}`);
  }

  getProductBySlug(slug: string): Observable<Product> {
    return this.http.get<Product>(`${this.apiUrl}/slug/${slug}`);
  }

  getProductsByCategory(categoryId: number, page: number, size: number): Observable<Product[]> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<Product[]>(`${this.apiUrl}/category/${categoryId}`, { params });
  }

  getFeaturedProducts(limit: number): Observable<Product[]> {
    const params = new HttpParams().set('limit', limit.toString());
    return this.http.get<Product[]>(`${this.apiUrl}/featured`, { params });
  }

  getDiscountedProducts(limit: number): Observable<Product[]> {
    const params = new HttpParams().set('limit', limit.toString());
    return this.http.get<Product[]>(`${this.apiUrl}/discounted`, { params });
  }

  getAllBrands(): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/brands`);
  }

  getPriceRange(searchRequest: ProductSearchRequest): Observable<PriceRange> {
    let params = new HttpParams();
    if (searchRequest.keyword) params = params.set('keyword', searchRequest.keyword);
    if (searchRequest.categoryId) params = params.set('categoryId', searchRequest.categoryId.toString());
    if (searchRequest.brand) params = params.set('brand', searchRequest.brand);
    if (searchRequest.inStock) params = params.set('inStock', searchRequest.inStock.toString());
    if (searchRequest.hasDiscount) params = params.set('hasDiscount', searchRequest.hasDiscount.toString());
    return this.http.get<PriceRange>(`${this.apiUrl}/price-range`, { params });
  }

  // Admin methods
  createProduct(request: ProductRequest): Observable<Product> {
    return this.http.post<Product>(this.apiUrl, request);
  }

  updateProduct(id: number, request: ProductRequest): Observable<Product> {
    return this.http.put<Product>(`${this.apiUrl}/${id}`, request);
  }

  deleteProduct(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  toggleProductStatus(id: number): Observable<Product> {
    return this.http.patch<Product>(`${this.apiUrl}/${id}/toggle-status`, {});
  }

  updateStock(id: number, quantity: number): Observable<Product> {
    const params = new HttpParams().set('quantity', quantity.toString());
    return this.http.patch<Product>(`${this.apiUrl}/${id}/stock`, {}, { params });
  }
}
